python/lsst/ts/salobj/domain.py

```diff
@@ -0,0 +1,250 @@
+    Environment variables:
+
+    * LSST_DDS_IP used to set the private_host field of samples
+      when writing them. Must be an integer. Optiona.
+      If absent then use a random positive number
+    * TS_IDL_DIR path to the ts_idl package. Required.
+
+    It is important to close a `Domain` when you are done with it, especially
+    in unit tests, because otherwise unreleased resources may cause problems.
+    To make this easier `Domain` is a context manager.
+
+    `Domain` contains a registry of `SalInfo` instances, so that it can close
+    them when the domain is closed. The registry uses weak references to avoid
+    circular dependencies (since SalInfo stores the domain). `SalInfo`
+    instances automatically register themselves with the provided domain.
+
+    To conserve resources you should have only one `Domain` per process.
+    Thus:
+
+    * `Controller` creates its own Domain as attribute ``domain``
+      and closes it when the controller is closed. To create a
+      `Remote` for a controller, use the domain from the controller, e.g.::
+
+        controller = Controller(name="Test", index=47)
+        try:
+            remote = Remote(domain=controller.domain, name="Test", index=47)
+            ...
+        finally:
+            await controller.close()
+
+        # or if you only want the objects for the duration of a function,
+        # e.g in a unit test:
+        async with Controller(name="Test", index=47) as controller:
+            remote = Remote(domain=controller.domain, name="Test", index=47)
+            ...
+
+    * If you are creating `Remote`s without a `Controller` then you must
+      create the Domain yourself and close it when you are finished, e.g.::
+
+        domain = salobj.Domain()
+        try:
+            remote = salobj.Remote(domain=domain, name="Test", index=47)
+            ...
+        finally:
+            await domain.close()
+
+        # or if you only want the objects for the duration of a function,
+        # e.g in a unit test:
+        async with salobj.Domain() as domain:
+            test_remote = salobj.Remote(domain=domain, name="Test", index=5)
+    """
+    def __init__(self):
+        self.participant = None
+
+        # accumulators for verifying that close is working
+        self.num_read_loops = 0
+        self.num_read_threads = 0
+
+        # set of _salinfo_set
+        self._salinfo_set = weakref.WeakSet()
+
+        host = os.environ.get("LSST_DDS_IP")
+        if host is None:
+            host = random.randint(1, MAX_RANDOM_HOST)
+        else:
+            try:
+                host = int(host)
+            except ValueError:
+                raise ValueError(f"Could not parse $LSST_DDS_IP={host} as an integer")
+        self.host = host
+        """Value for the private_host field of output samples."""
+
+        self.origin = os.getpid()
+        """Value for the private_origin field of output samples."""
+
+        idl_dir = os.environ.get("TS_IDL_DIR")
```

Comment:
@r-owen , do you think you can change the strategy to get the idl files? As from our previous conversation, I’m not a big fan of using environment variables… I have two proposed suggestions, in order of preference:

1 - Remove ts_idl and add an “idl” directory at the top level of salobj where the idl and qos files are stored. If this is the case, "Test" idl files could even be part of the package, along with the qos files. 
2 - Keep ts_idl but add a python package that defines the paths, then import the path from the idl package. Something like:

```
from lsst.ts.idl import idl_dir
```

and idl_dir could be defined like:

```
idl_dir =  pathlib.Path(__file__).resolve().parents[4]
```

or something like that.

Reply:
I think all languages need access to the QoS file, which means putting it in ts_salobj or using a python path is not a good solution for that. I should check to see what other languages are doing and consider using that instead of putting it in ts_idl.

I am uneasy about putting the IDL files in ts_salobj, because they change on a different time scale than ts_salobj itself. This may make XML and salobj updates a bit complicated.

Personally I think environment variables are better than relying on Python paths for getting this kind of data. In theory other languages might also want to get to these IDL files (though I think it unlikely), or we might want to put other kinds of SAL libraries in this package (which I think would be great, but then the name isn't ideal). Neither use case can use the python `module.__file__` trick.

Reply:
I asked @dmills-lsst about the qos file `DDS_DefaultQoS_All.xml` and he said that he plans to transition all CSCs to use that file, and plans to use environment variable `LSST_DDS_QOS` to point to it. I am a bit worried about where the file should live for ts_salobj but I should probably go along with this plan and modify ts_salobj to use this. I'm not sure how ts_salobj can safely find the shared common QoS file except by using an environment variable.

Reply:
I have updated ts_idl to include the following:
* get_idl_dir() a function that returns the root dir of the package
* get_qos_path() a function that returns the path to the QoS file; for now it points to a copy in ts_idl but that could change if @dmills-lsst feels strongly enough about it.
* lsst.ts.idl.enums: a set of enum files for various CSCs. For now I have enums for ATDome, ATMCS, Script and ScriptQueue. I will add more as needed. They are manually generated for now and I suspect we'll just stick to that, but if we can generate them automatically (preferably without changing the names for the classes and enum constants) then that would be even better.